File: src/main/scala/com/sparkTutorial/advanced/accumulator/StackOverFlowSurveyFollowUp.scala
package com.sparkTutorial.advanced.accumulator

import com.sparkTutorial.commons.Utils
import org.apache.log4j.{Level, Logger}
import org.apache.spark.{SparkConf, SparkContext}

object StackOverFlowSurveyFollowUp {

  def main(args: Array[String]) {

    Logger.getLogger("org").setLevel(Level.ERROR)
    val conf = new SparkConf().setAppName("StackOverFlowSurvey").setMaster("local[1]")
    val sparkContext = new SparkContext(conf)

    val total = sparkContext.longAccumulator
    val missingSalaryMidPoint = sparkContext.longAccumulator
    val processedBytes = sparkContext.longAccumulator

    val responseRDD = sparkContext.textFile("in/2016-stack-overflow-survey-responses.csv")

    val responseFromCanada = responseRDD.filter(response => {

      processedBytes.add(response.getBytes().length)
      val splits = response.split(Utils.COMMA_DELIMITER, -1)
      total.add(1)

      if (splits(14).isEmpty) {
        missingSalaryMidPoint.add(1)
      }
      splits(2) == "Canada"

    })

    println("Count of responses from Canada: " + responseFromCanada.count())
    println("Number of bytes processed: " + processedBytes.value)
    println("Total count of responses: " + total.value)
    println("Count of responses missing salary middle point: " + missingSalaryMidPoint.value)
  }
}

File: advanced/accumulator/StackOverFlowSurveyFollowUp.py
from pyspark import SparkContext
from commons.Utils import Utils

def filterResponseFromCanada(response, total, missingSalaryMidPoint, processedBytes):
    processedBytes.add(len(response.encode('utf-8')))
    splits = Utils.COMMA_DELIMITER.split(response)
    total.add(1)
    if not splits[14]:
        missingSalaryMidPoint.add(1)
    return splits[2] == "Canada"

if __name__ == "__main__":
    sc = SparkContext("local", "StackOverFlowSurvey")
    sc.setLogLevel("ERROR")

    total = sc.accumulator(0)
    missingSalaryMidPoint = sc.accumulator(0)
    processedBytes = sc.accumulator(0)

    responseRDD = sc.textFile("in/2016-stack-overflow-survey-responses.csv")

    responseFromCanada = responseRDD.filter(lambda response: \
        filterResponseFromCanada(response, total, missingSalaryMidPoint, processedBytes))

    print("Count of responses from Canada: {}".format(responseFromCanada.count()))
    print("Number of bytes processed: {}".format(processedBytes.value))
    print("Total count of responses: {}".format(total.value))
    print("Count of responses missing salary middle point: {}".format(missingSalaryMidPoint.value))
